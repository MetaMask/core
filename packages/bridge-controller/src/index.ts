export { BridgeController } from './bridge-controller';

export type {
  ChainConfiguration,
  L1GasFees,
  SolanaFees,
  QuoteMetadata,
  BridgeToken,
  GasMultiplierByChainId,
  FeatureFlagResponse,
  BridgeAsset,
  GenericQuoteRequest,
  Protocol,
  TokenAmountValues,
  Step,
  RefuelData,
  Quote,
  QuoteResponse,
  FeeData,
  TxData,
  BridgeFeatureFlags,
  BridgeControllerState,
  BridgeControllerAction,
  BridgeControllerActions,
  BridgeControllerEvents,
  BridgeControllerMessenger,
} from './types';

export {
  AssetType,
  SortOrder,
  BridgeFlag,
  ActionTypes,
  ChainId,
  BridgeFeatureFlagsKey,
  RequestStatus,
  BridgeUserAction,
  BridgeBackgroundAction,
  FeeType,
} from './types';

export {
  ALLOWED_BRIDGE_CHAIN_IDS,
  BridgeClientId,
  BRIDGE_CONTROLLER_NAME,
  BRIDGE_QUOTE_MAX_ETA_SECONDS,
  BRIDGE_QUOTE_MAX_RETURN_DIFFERENCE_PERCENTAGE,
  BRIDGE_PREFERRED_GAS_ESTIMATE,
  BRIDGE_DEFAULT_SLIPPAGE,
  BRIDGE_MM_FEE_RATE,
  REFRESH_INTERVAL_MS,
  DEFAULT_MAX_REFRESH_COUNT,
  DEFAULT_BRIDGE_CONTROLLER_STATE,
  METABRIDGE_CHAIN_TO_ADDRESS_MAP,
  BRIDGE_DEV_API_BASE_URL,
  BRIDGE_PROD_API_BASE_URL,
} from './constants/bridge';

export type { AllowedBridgeChainIds } from './constants/bridge';

export {
  /**
   * @deprecated This type should not be used. Use {@link BridgeAsset} instead.
   */
  type SwapsTokenObject,
  /**
   * @deprecated This map should not be used. Use getNativeAssetForChainId" } instead.
   */
  SWAPS_CHAINID_DEFAULT_TOKEN_MAP,
} from './constants/tokens';

export { SWAPS_API_V2_BASE_URL } from './constants/swaps';

export {
  getEthUsdtResetData,
  isEthUsdt,
  isNativeAddress,
  isSolanaChainId,
  getNativeAssetForChainId,
  getDefaultBridgeControllerState,
} from './utils/bridge';

export { isValidQuoteRequest, formatEtaInMinutes } from './utils/quote';

export { calcLatestSrcBalance } from './utils/balance';

export { fetchBridgeTokens } from './utils/fetch';

export {
  formatChainIdToCaip,
  formatChainIdToHex,
  formatAddressToCaipReference,
} from './utils/caip-formatters';

export {
  selectBridgeQuotes,
  type BridgeAppState,
  selectExchangeRateByChainIdAndAddress,
  /**
   * Returns whether a quote is expired
   *
   * @param state The state of the bridge controller and its dependency controllers
   * @param currentTimeInMs The current timestamp in milliseconds (e.g. `Date.now()`)
   * @returns Whether the quote is expired
   */
  selectIsQuoteExpired,
} from './selectors';
