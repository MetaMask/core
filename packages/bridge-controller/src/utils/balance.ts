import { getAddress } from '@ethersproject/address';
import type { BigNumber } from '@ethersproject/bignumber';
import { Contract } from '@ethersproject/contracts';
import { Web3Provider } from '@ethersproject/providers';
import { abiERC20 } from '@metamask/metamask-eth-abis';
import type { Provider } from '@metamask/network-controller';
import type { Hex } from '@metamask/utils';

import { isNativeAddress } from './bridge';

export const fetchTokenBalance = async (
  address: string,
  userAddress: string,
  provider: Provider,
): Promise<BigNumber | undefined> => {
  const ethersProvider = new Web3Provider(provider);
  const tokenContract = new Contract(address, abiERC20, ethersProvider);
  const tokenBalancePromise =
    typeof tokenContract?.balanceOf === 'function'
      ? tokenContract.balanceOf(userAddress)
      : Promise.resolve(undefined);
  return await tokenBalancePromise;
};

export const calcLatestSrcBalance = async (
  provider: Provider,
  selectedAddress: string,
  tokenAddress: string,
  chainId: Hex,
): Promise<BigNumber | undefined> => {
  if (tokenAddress && chainId) {
    if (isNativeAddress(tokenAddress)) {
      const ethersProvider = new Web3Provider(provider);
      return await ethersProvider.getBalance(getAddress(selectedAddress));
    }
    return await fetchTokenBalance(tokenAddress, selectedAddress, provider);
  }
  return undefined;
};

export const hasSufficientBalance = async (
  provider: Provider,
  selectedAddress: string,
  tokenAddress: string,
  fromTokenAmount: string,
  chainId: Hex,
) => {
  const srcTokenBalance = await calcLatestSrcBalance(
    provider,
    selectedAddress,
    tokenAddress,
    chainId,
  );

  return srcTokenBalance ? srcTokenBalance.gte(fromTokenAmount) : false;
};
