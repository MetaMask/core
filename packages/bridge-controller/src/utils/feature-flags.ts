import { formatChainIdToCaip } from './caip-formatters';
import { validateFeatureFlagsResponse } from './validators';
import { DEFAULT_FEATURE_FLAG_CONFIG } from '../constants/bridge';
import type {
  FeatureFlagsPlatformConfig,
  ChainConfiguration,
  BridgeControllerMessenger,
} from '../types';

export const formatFeatureFlags = (
  bridgeFeatureFlags: FeatureFlagsPlatformConfig,
) => {
  const getChainsObj = (chains: Record<string, ChainConfiguration>) =>
    Object.entries(chains).reduce(
      (acc, [chainId, value]) => ({
        ...acc,
        [formatChainIdToCaip(chainId)]: value,
      }),
      {},
    );

  return {
    ...bridgeFeatureFlags,
    chains: getChainsObj(bridgeFeatureFlags.chains),
  };
};

export const processFeatureFlags = (
  bridgeFeatureFlags: unknown,
): FeatureFlagsPlatformConfig => {
  if (validateFeatureFlagsResponse(bridgeFeatureFlags)) {
    return formatFeatureFlags(bridgeFeatureFlags);
  }
  return DEFAULT_FEATURE_FLAG_CONFIG;
};

/**
 * Gets the bridge feature flags from the remote feature flag controller
 *
 * @param messenger - The messenger instance
 * @returns The bridge feature flags
 */
export function getBridgeFeatureFlags(
  messenger: BridgeControllerMessenger,
): FeatureFlagsPlatformConfig {
  // This will return the bridgeConfig for the current platform even without specifying the platform
  const remoteFeatureFlagControllerState = messenger.call(
    'RemoteFeatureFlagController:getState',
  );
  const rawBridgeConfig =
    remoteFeatureFlagControllerState?.remoteFeatureFlags?.bridgeConfig;

  return processFeatureFlags(rawBridgeConfig);
}
