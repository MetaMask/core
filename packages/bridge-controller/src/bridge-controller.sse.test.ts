/* eslint-disable jest/no-restricted-matchers */

import { SolScope } from '@metamask/keyring-api';
import * as eventSource from '@microsoft/fetch-event-source';

import { BridgeController } from './bridge-controller';
import {
  BridgeClientId,
  BRIDGE_PROD_API_BASE_URL,
  DEFAULT_BRIDGE_CONTROLLER_STATE,
} from './constants/bridge';
import { ChainId, type BridgeControllerMessenger } from './types';
import * as balanceUtils from './utils/balance';
import * as featureFlagUtils from './utils/feature-flags';
import * as fetchUtils from './utils/fetch';
import { flushPromises } from '../../../tests/helpers';
import { handleFetch } from '../../controller-utils/src';
import mockBridgeQuotesNativeErc20Eth from '../tests/mock-quotes-native-erc20-eth.json';
import mockBridgeQuotesNativeErc20 from '../tests/mock-quotes-native-erc20.json';

const messengerMock = {
  call: jest.fn(),
  registerActionHandler: jest.fn(),
  registerInitialEventPayload: jest.fn(),
  publish: jest.fn(),
} as unknown as jest.Mocked<BridgeControllerMessenger>;

const getLayer1GasFeeMock = jest.fn();
const mockFetchFn = handleFetch;
const trackMetaMetricsFn = jest.fn();
let fetchAssetPricesSpy: jest.SpyInstance;

describe('BridgeController SSE tests', function () {
  let bridgeController: BridgeController;

  beforeEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
    jest.resetAllMocks();

    fetchAssetPricesSpy = jest
      .spyOn(fetchUtils, 'fetchAssetPrices')
      .mockResolvedValue({
        'eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984': {
          usd: '100',
        },
      });

    getLayer1GasFeeMock.mockResolvedValue('0x1');
    bridgeController = new BridgeController({
      messenger: messengerMock,
      getLayer1GasFee: getLayer1GasFeeMock,
      clientId: BridgeClientId.EXTENSION,
      fetchFn: mockFetchFn,
      trackMetaMetricsFn,
      clientVersion: '1.0.0',
    });
    messengerMock.call.mockReturnValue({
      address: '0x123',
      provider: jest.fn(),
      selectedNetworkClientId: 'selectedNetworkClientId',
      currencyRates: {},
      marketData: {},
      conversionRates: {},
    } as never);
    const defaultFlags = {
      minimumVersion: '0.0.0',
      maxRefreshCount: 5,
      refreshRate: 30000,
      support: true,
      sseEnabled: true,
      chains: {
        '10': { isActiveSrc: true, isActiveDest: false },
        '534352': { isActiveSrc: true, isActiveDest: false },
        '137': { isActiveSrc: false, isActiveDest: true },
        '42161': { isActiveSrc: false, isActiveDest: true },
        [ChainId.SOLANA]: {
          isActiveSrc: true,
          isActiveDest: true,
        },
      },
    };
    jest
      .spyOn(featureFlagUtils, 'getBridgeFeatureFlags')
      .mockReturnValue(defaultFlags);
  });

  const metricsContext = {
    token_symbol_source: 'ETH',
    token_symbol_destination: 'USDC',
    usd_amount_source: 100,
    stx_enabled: true,
    security_warnings: [],
    warnings: [],
  };

  it('updateBridgeQuoteRequestParams should trigger quote polling if request is valid', async function () {
    jest.useFakeTimers();
    const stopAllPollingSpy = jest.spyOn(bridgeController, 'stopAllPolling');
    const startPollingSpy = jest.spyOn(bridgeController, 'startPolling');
    const hasSufficientBalanceSpy = jest
      .spyOn(balanceUtils, 'hasSufficientBalance')
      .mockResolvedValue(true);

    const fetchBridgeQuotesSpy = jest.spyOn(
      fetchUtils,
      'fetchBridgeQuoteStream',
    );

    const consoleLogSpy = jest.spyOn(console, 'log');

    jest
      .spyOn(eventSource, 'fetchEventSource')
      .mockImplementationOnce(async (url, { onopen, onmessage, onclose }) => {
        setTimeout(() => {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          onopen?.({ ok: true } as Response);
        }, 1000);

        setTimeout(() => {
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20[0]),
            event: 'quote',
            id: '1',
          });
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20[1]),
            event: 'quote',
            id: '2',
          });
          onclose?.();
        }, 4000);
      })
      .mockImplementationOnce(async (url, { onopen, onmessage }) => {
        setTimeout(() => {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          onopen?.({ ok: true } as Response);
        }, 1000);

        setTimeout(() => {
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20Eth[0]),
            event: 'quote',
            id: '1',
          });
        }, 9000);
        await Promise.resolve();

        setTimeout(() => {
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20Eth[1]),
            event: 'quote',
            id: '2',
          });
        }, 2000);
      })
      .mockImplementationOnce(async (url, { onopen, onmessage, onerror }) => {
        setTimeout(() => {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          onopen?.({ ok: true } as Response);
        }, 1000);

        onerror?.('Network error');
      })
      .mockImplementationOnce(async (url, { onopen, onmessage }) => {
        setTimeout(() => {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          onopen?.({ ok: true } as Response);
        }, 1000);

        setTimeout(() => {
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20[0]),
            event: 'quote',
            id: '1',
          });
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20[1]),
            event: 'quote',
            id: '2',
          });
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20[0]),
            event: 'quote',
            id: '3',
          });
          onmessage?.({
            data: JSON.stringify(mockBridgeQuotesNativeErc20[1]),
            event: 'quote',
            id: '4',
          });
        }, 10000);
      });

    const quoteParams = {
      srcChainId: '0x1',
      destChainId: SolScope.Mainnet,
      srcTokenAddress: '0x0000000000000000000000000000000000000000',
      destTokenAddress: '123d1',
      srcTokenAmount: '1000000000000000000',
      slippage: 0.5,
      walletAddress: '0x30E8ccaD5A980BDF30447f8c2C48e70989D9d294',
      destWalletAddress: 'SolanaWalletAddres1234',
    };
    const quoteRequest = {
      ...quoteParams,
    };
    await bridgeController.updateBridgeQuoteRequestParams(
      quoteParams,
      metricsContext,
    );

    // Before polling starts
    expect(stopAllPollingSpy).toHaveBeenCalledTimes(1);
    expect(startPollingSpy).toHaveBeenCalledTimes(1);
    expect(hasSufficientBalanceSpy).toHaveBeenCalledTimes(1);
    expect(startPollingSpy).toHaveBeenCalledWith({
      networkClientId: 'selectedNetworkClientId',
      updatedQuoteRequest: {
        ...quoteRequest,
        insufficientBal: false,
      },
      context: metricsContext,
    });
    expect(fetchAssetPricesSpy).toHaveBeenCalledTimes(1);
    expect(bridgeController.state).toStrictEqual(
      expect.objectContaining({
        quoteRequest: {
          ...quoteRequest,
          walletAddress: '0x30E8ccaD5A980BDF30447f8c2C48e70989D9d294',
        },
        quotes: DEFAULT_BRIDGE_CONTROLLER_STATE.quotes,
        quotesLastFetched: DEFAULT_BRIDGE_CONTROLLER_STATE.quotesLastFetched,
        quotesLoadingStatus:
          DEFAULT_BRIDGE_CONTROLLER_STATE.quotesLoadingStatus,
      }),
    );

    // Loading state
    jest.advanceTimersByTime(1000);
    expect(fetchBridgeQuotesSpy).toHaveBeenCalledTimes(1);
    expect(fetchBridgeQuotesSpy).toHaveBeenCalledWith(
      mockFetchFn,
      {
        ...quoteRequest,
        insufficientBal: false,
      },
      expect.any(AbortSignal),
      BridgeClientId.EXTENSION,
      BRIDGE_PROD_API_BASE_URL,
      {
        onOpen: expect.any(Function),
        onValidationFailure: expect.any(Function),
        onValidQuoteReceived: expect.any(Function),
      },
      '1.0.0',
    );
    expect(bridgeController.state.quotesLastFetched).toBeNull();
    expect(bridgeController.state).toMatchSnapshot();

    // After first fetch
    jest.advanceTimersByTime(5000);
    await flushPromises();
    expect(consoleLogSpy.mock.calls).toMatchInlineSnapshot(`Array []`);
    expect(bridgeController.state).toStrictEqual(
      expect.objectContaining({
        quotesInitialLoadTime: 6000,
        quoteRequest: { ...quoteRequest, insufficientBal: false },
        quotes: mockBridgeQuotesNativeErc20.map((quote) => ({
          ...quote,
          l1GasFeesInHexWei: '0x1',
        })),
        quotesRefreshCount: 1,
        quotesLoadingStatus: 1,
      }),
    );
    const firstFetchTime = bridgeController.state.quotesLastFetched;
    expect(firstFetchTime).toBeGreaterThan(0);
    jest.advanceTimersToNextTimer();

    // After 2nd fetch's first server event
    jest.advanceTimersByTime(11000);
    await flushPromises();
    expect(bridgeController.state.quotes).toHaveLength(1);
    expect(bridgeController.state).toStrictEqual(
      expect.objectContaining({
        quotesInitialLoadTime: 6000,
        quoteRequest: { ...quoteRequest, insufficientBal: false },
        quotes: [mockBridgeQuotesNativeErc20Eth[0]],
        quotesLoadingStatus: 1,
        quoteFetchError: null,
        quotesRefreshCount: 2,
      }),
    );
    expect(fetchBridgeQuotesSpy).toHaveBeenCalledTimes(2);
    const secondFetchTime = bridgeController.state.quotesLastFetched;
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(secondFetchTime).toBeGreaterThan(firstFetchTime!);

    // After 2nd server event
    jest.advanceTimersByTime(2000);
    await flushPromises();
    expect(bridgeController.state.quotes).toHaveLength(2);
    expect(bridgeController.state).toStrictEqual(
      expect.objectContaining({
        quotesInitialLoadTime: 6000,
        quoteRequest: { ...quoteRequest, insufficientBal: false },
        quotes: mockBridgeQuotesNativeErc20Eth,
        quotesLoadingStatus: 1,
        quoteFetchError: null,
        quotesRefreshCount: 2,
      }),
    );
    expect(fetchBridgeQuotesSpy).toHaveBeenCalledTimes(2);
    const secondFetchNextMessageTime = bridgeController.state.quotesLastFetched;
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(secondFetchNextMessageTime).toBe(secondFetchTime!);

    // After 3nd fetch throws an error
    jest.advanceTimersToNextTimer();
    await flushPromises();
    expect(fetchBridgeQuotesSpy).toHaveBeenCalledTimes(3);
    expect(bridgeController.state).toStrictEqual(
      expect.objectContaining({
        quotesInitialLoadTime: 6000,
        quoteRequest: { ...quoteRequest, insufficientBal: false },
        quotes: [],
        quotesLoadingStatus: 2,
        quoteFetchError: 'Network error',
        quotesRefreshCount: 3,
      }),
    );
    const thirdFetchTime = bridgeController.state.quotesLastFetched;
    expect(
      thirdFetchTime,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ).toBeGreaterThan(secondFetchTime!);
    expect(consoleLogSpy).toHaveBeenCalledWith(
      'Failed to stream bridge quotes',
      new Error('Network error'),
    );
    expect(consoleLogSpy.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "Failed to stream bridge quotes",
          [Error: Network error],
        ],
      ]
    `);

    // Incoming request update aborts current polling
    jest.advanceTimersByTime(10000);
    await flushPromises();
    await bridgeController.updateBridgeQuoteRequestParams(
      { ...quoteRequest, srcTokenAmount: '10', insufficientBal: false },
      {
        stx_enabled: true,
        token_symbol_source: 'ETH',
        token_symbol_destination: 'USDC',
        security_warnings: [],
      },
    );
    await flushPromises();
    expect(fetchBridgeQuotesSpy).toHaveBeenCalledTimes(3);
    expect(bridgeController.state).toMatchSnapshot();

    // Next fetch succeeds
    jest.advanceTimersByTime(15000);
    await flushPromises();
    expect(fetchBridgeQuotesSpy).toHaveBeenCalledTimes(4);
    const { quotesLastFetched, quotes, ...stateWithoutTimestamp } =
      bridgeController.state;

    expect(stateWithoutTimestamp).toMatchSnapshot();
    expect(quotes).toStrictEqual(
      [...mockBridgeQuotesNativeErc20, ...mockBridgeQuotesNativeErc20].map(
        (quote) => ({
          ...quote,
          l1GasFeesInHexWei: '0x1',
        }),
      ),
    );
    expect(
      quotesLastFetched,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ).toBeGreaterThan(thirdFetchTime!);

    expect(consoleLogSpy).toHaveBeenCalledTimes(1);
    expect(hasSufficientBalanceSpy).toHaveBeenCalledTimes(1);
    expect(getLayer1GasFeeMock).toHaveBeenCalledTimes(6);
    expect(trackMetaMetricsFn).toHaveBeenCalledTimes(9);
    expect(trackMetaMetricsFn.mock.calls).toMatchSnapshot();
  });
});
