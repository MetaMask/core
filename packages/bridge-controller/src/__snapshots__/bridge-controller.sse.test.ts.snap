// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BridgeController SSE should publish validation failures 3`] = `
Array [
  Array [
    "Unified SwapBridge Quotes Failed Validation",
    Object {
      "action_type": "swapbridge-v1",
      "chain_id_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
      "chain_id_source": "eip155:1",
      "failures": Array [
        "lifi|trade",
        "lifi|trade.chainId",
        "lifi|trade.to",
        "lifi|trade.from",
        "lifi|trade.value",
        "lifi|trade.data",
        "lifi|trade.gasLimit",
        "lifi|trade.unsignedPsbtBase64",
        "lifi|trade.inputsToSign",
      ],
      "refresh_count": 1,
      "token_address_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
      "token_address_source": "eip155:1/slip44:60",
    },
  ],
  Array [
    "Unified SwapBridge Quotes Failed Validation",
    Object {
      "action_type": "swapbridge-v1",
      "chain_id_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
      "chain_id_source": "eip155:1",
      "failures": Array [
        "unknown|quote",
      ],
      "refresh_count": 2,
      "token_address_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
      "token_address_source": "eip155:1/slip44:60",
    },
  ],
]
`;

exports[`BridgeController SSE should replace all stale quotes after a refresh and first quote is received 1`] = `
Array [
  "Unified SwapBridge Quotes Requested",
  Object {
    "action_type": "swapbridge-v1",
    "chain_id_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
    "chain_id_source": "eip155:1",
    "custom_slippage": true,
    "has_sufficient_funds": true,
    "is_hardware_wallet": false,
    "security_warnings": Array [],
    "slippage_limit": 0.5,
    "stx_enabled": true,
    "swap_type": "crosschain",
    "token_address_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
    "token_address_source": "eip155:1/slip44:60",
    "token_symbol_destination": "USDC",
    "token_symbol_source": "ETH",
    "usd_amount_source": 100,
    "warnings": Array [],
  },
]
`;

exports[`BridgeController SSE should reset and refetch quotes after quote request is changed 1`] = `
Array [
  Array [
    "Unified SwapBridge Quotes Requested",
    Object {
      "action_type": "swapbridge-v1",
      "chain_id_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
      "chain_id_source": "eip155:1",
      "custom_slippage": true,
      "has_sufficient_funds": true,
      "is_hardware_wallet": false,
      "security_warnings": Array [],
      "slippage_limit": 0.5,
      "stx_enabled": true,
      "swap_type": "crosschain",
      "token_address_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
      "token_address_source": "eip155:1/slip44:60",
      "token_symbol_destination": "USDC",
      "token_symbol_source": "ETH",
    },
  ],
]
`;

exports[`BridgeController SSE should reset quotes list if quote refresh fails 2`] = `
Array [
  Array [
    "Unified SwapBridge Quotes Requested",
    Object {
      "action_type": "swapbridge-v1",
      "chain_id_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
      "chain_id_source": "eip155:1",
      "custom_slippage": true,
      "has_sufficient_funds": true,
      "is_hardware_wallet": false,
      "security_warnings": Array [],
      "slippage_limit": 0.5,
      "stx_enabled": true,
      "swap_type": "crosschain",
      "token_address_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
      "token_address_source": "eip155:1/slip44:60",
      "token_symbol_destination": "USDC",
      "token_symbol_source": "ETH",
      "usd_amount_source": 100,
      "warnings": Array [],
    },
  ],
  Array [
    "Unified SwapBridge Quotes Error",
    Object {
      "action_type": "swapbridge-v1",
      "chain_id_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
      "chain_id_source": "eip155:1",
      "custom_slippage": true,
      "error_message": "Network error",
      "has_sufficient_funds": true,
      "is_hardware_wallet": false,
      "security_warnings": Array [],
      "slippage_limit": 0.5,
      "stx_enabled": true,
      "swap_type": "crosschain",
      "token_address_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
      "token_address_source": "eip155:1/slip44:60",
      "token_symbol_destination": "USDC",
      "token_symbol_source": "ETH",
      "usd_amount_source": 100,
      "warnings": Array [],
    },
  ],
]
`;

exports[`BridgeController SSE should trigger quote polling if request is valid 1`] = `
Object {
  "assetExchangeRates": Object {
    "eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984": Object {
      "exchangeRate": undefined,
      "usdExchangeRate": "100",
    },
  },
  "minimumBalanceForRentExemptionInLamports": "0",
  "quoteFetchError": null,
  "quoteRequest": Object {
    "destChainId": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
    "destTokenAddress": "123d1",
    "destWalletAddress": "SolanaWalletAddres1234",
    "insufficientBal": false,
    "slippage": 0.5,
    "srcChainId": "0x1",
    "srcTokenAddress": "0x0000000000000000000000000000000000000000",
    "srcTokenAmount": "1000000000000000000",
    "walletAddress": "0x30E8ccaD5A980BDF30447f8c2C48e70989D9d294",
  },
  "quotes": Array [],
  "quotesInitialLoadTime": null,
  "quotesLoadingStatus": 0,
  "quotesRefreshCount": 0,
}
`;

exports[`BridgeController SSE should trigger quote polling if request is valid 2`] = `
Array [
  Array [
    "Unified SwapBridge Input Changed",
    Object {
      "action_type": "swapbridge-v1",
      "input": "chain_source",
      "input_value": "eip155:1",
    },
  ],
  Array [
    "Unified SwapBridge Input Changed",
    Object {
      "action_type": "swapbridge-v1",
      "input": "chain_destination",
      "input_value": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
    },
  ],
  Array [
    "Unified SwapBridge Input Changed",
    Object {
      "action_type": "swapbridge-v1",
      "input": "token_destination",
      "input_value": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
    },
  ],
  Array [
    "Unified SwapBridge Input Changed",
    Object {
      "action_type": "swapbridge-v1",
      "input": "slippage",
      "input_value": 0.5,
    },
  ],
  Array [
    "Unified SwapBridge Quotes Requested",
    Object {
      "action_type": "swapbridge-v1",
      "chain_id_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
      "chain_id_source": "eip155:1",
      "custom_slippage": true,
      "has_sufficient_funds": true,
      "is_hardware_wallet": false,
      "security_warnings": Array [],
      "slippage_limit": 0.5,
      "stx_enabled": true,
      "swap_type": "crosschain",
      "token_address_destination": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp/token:123d1",
      "token_address_source": "eip155:1/slip44:60",
      "token_symbol_destination": "USDC",
      "token_symbol_source": "ETH",
      "usd_amount_source": 100,
      "warnings": Array [],
    },
  ],
]
`;
